src/index.ts

```diff
@@ -0,0 +1,109 @@
+import os from 'os';
+import {
+  CosmiconfigResult,
+  ExplorerOptions,
+  Loaders,
+  LoaderEntry,
+  SyncLoader,
+  AsyncLoader,
+} from './types';
+import { createExplorer } from './createExplorer';
+import * as loaders from './loaders';
+
+function identity<T>(x: T): T {
+  return x;
+}
+
+// TODO: can it be undefined to disable a loader?
+interface RawLoaders {
+  [key: string]: LoaderEntry | SyncLoader | AsyncLoader;
+}
+
+function normalizeLoaders(rawLoaders?: RawLoaders): Loaders {
+  const defaultLoaders: Loaders = {
+    '.js': { sync: loaders.loadJs, async: loaders.loadJs },
+    '.json': { sync: loaders.loadJson, async: loaders.loadJson },
+    '.yaml': { sync: loaders.loadYaml, async: loaders.loadYaml },
+    '.yml': { sync: loaders.loadYaml, async: loaders.loadYaml },
+    noExt: { sync: loaders.loadYaml, async: loaders.loadYaml },
+  };
+
+  if (!rawLoaders) {
+    return defaultLoaders;
+  }
+
+  const result = Object.keys(rawLoaders).reduce(
+    (result: Loaders, ext: string): Loaders => {
+      const entry = rawLoaders && rawLoaders[ext];
+
+      if (typeof entry === 'function') {
+        // TODO: sync loader can be async here
+        // TODO: direct remove function support and specify both async and sync?
+        const sync: SyncLoader = entry;
+        const async: AsyncLoader = entry;
+
```

Comment:
Currently a user can provide an async function to to `sync`, which will return unexpected result. Also, as far as I know, the loaders option cannot be properly typed. I think we should drop the function support and only support the `{ sync: () => 'syncLoader', async: async () => 'asyncLoader' }` option object.